#17230 API design: Fleet in your calendar (#17552)
fleetdm/fleet: docs/REST API/rest-api.md

    }
  },
  "integrations": {
    "jira": null,

>     "google_calendar": [

Reviewer: is there a reason this is an array and not an object? @rachaelshaw @getvictor

Reviewer 2: We will support multiple configs in the future. For example, one company acquires another, so they are on different Google Workspace accounts and with different domains.

Reviewer 3: What will "not configured" look like? `[]`? `null`?

Reviewer 2: For global configs it will be `[]` or `null` or `[null]`, for team config it will be `null`

Reviewer: thank you!

      {

>         "email": "name@example.com",
>         "domain": "example.com",
>         "private_key": "abc123"

Author: ```suggestion
        "domain": "example.com",
        "private_key_json": {
           "type": "service_account",
           "project_id": "fleet-in-your-calendar",
           "private_key_id": "<private key id>",
           "private_key": "-----BEGIN PRIVATE KEY-----\n<private key>\n-----END PRIVATE KEY-----\n",
           "client_email": "fleet-calendar-events@fleet-in-your-calendar.iam.gserviceaccount.com",
           "client_id": "<client id>",
           "auth_uri": "https://accounts.google.com/o/oauth2/auth",
           "token_uri": "https://oauth2.googleapis.com/token",
           "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
           "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/fleet-calendar-events%40fleet-in-your-calendar.iam.gserviceaccount.com",
           "universe_domain": "googleapis.com"
         }
```

| email                             | string  | body  | _integrations.zendesk[] settings_. The Zendesk user email to use for this Zendesk integration. |
| api_token                         | string  | body  | _integrations.zendesk[] settings_. The Zendesk API token to use for this Zendesk integration. |
| group_id                          | integer | body  | _integrations.zendesk[] settings_. The Zendesk group id to use for this integration. Zendesk tickets will be created in this group. |

> | email                             | string  | body  | _integrations.google_calendar[] settings_. The email address for the Google Workspace service account to be used for this calendar integration. |

Author: ```suggestion
```

> | email                             | string  | body  | _integrations.google_calendar[] settings_. The email address for the Google Workspace service account to be used for this calendar integration. |

Reviewer 3: @rachaelshaw  looks like there's no longer and `email` field [here](https://github.com/fleetdm/fleet/blob/5eb8cf618e0c321bde3b2883398c422eaaefd513/frontend/pages/admin/IntegrationsPage/cards/Calendars/Calendars.tsx#L58) unless I'm wrong @RachelElysia

| domain                            | string  | body  | _integrations.google_calendar[] settings_. The domain for the Google Workspace service account to be used for this calendar integration. |

> | private_key                       | string  | body  | _integrations.google_calendar[] settings_. The private key for the Google Workspace service account to be used for this calendar integration. |

Author: ```suggestion
| private_key_json                       | TODO  | body  | _integrations.google_calendar[] settings_. The private key JSON downloaded when generating service account API key to be used for this calendar integration. |
```

Reviewer 4: @rachaelshaw and @getvictor when you get the chance, can you please follow up to the above? Thanks!

Hoping to get these doc changes merged ASAP

Author: ^^ cc @RachelElysia in case you know the answer (since Victor's OOO)

Reviewer 3: and it can be `null`

        "host_batch_size": 0
      }
    },

>     "integrations": {

Reviewer 3: @rachaelshaw @getvictor What API field does [this form field](https://www.figma.com/file/p81nWodxL04YD7iNyr9xVa/%2317230-Fleet-in-your-calendar?type=design&node-id=513%3A9054&mode=design&t=7Nyma5ovXM6NPVWj-1) translate to? Are we missing a field here?

>       "google_calendar": {

Reviewer 3: @rachaelshaw @getvictor What API field does the "[Resolution webhook url"](https://www.figma.com/file/p81nWodxL04YD7iNyr9xVa/%2317230-Fleet-in-your-calendar?type=design&node-id=513%3A9265&mode=dev) form field translate to? Are we missing a field here?

Reviewer 3: Maybe the below `email` field should be `resolution_webhook_url`?

Reviewer 3: So when setting, `policies` subfields can be `name` or `id`, but the API always returns `policy_ids: [1,2,3]` right?

Author: Can we talk about this more @ standup? I want to understand why a `policy_ids` parameter won't work here. (It's more consistent w/ other parts of the API if we're able to accept an array of numbers here instead of an array of objects.)

Author: > It is `webhook_url`.

Looks like we call it `destination_url` for failing policy automations, but definitely agree `webhook_url` is better here, since the calendar integration could potentially have other options in the future besides triggering a webhook.

        "enable_calendar_events": true,
        "email": "name@example.com",

>         "policy_ids": [1, 2, 3]

Reviewer 2: @jacobshandling This is out of date. It is:
```
  "policies": [{"name": "policy 1", "id": 1}, {"name": "policy 2", "id": 2}]
```

The names are validated and ids are set when config is saved. When saving, only names are needed.

This means the names and/or ids may not match the actual policies you're seeing.
- name may have been renamed
- id may be gone (deleted)

Try to match on names, so that we are consistent with how we will set configs via fleetctl apply/gitops.

Reviewer: @rachaelshaw can we update this according to @getvictor 's comment so this PR is the source of truth?

Reviewer 2: It is possible to support both. Maybe give priority to names. If name is blank, then use the id.
